Block bulk request's index metadata retrieval until cluster state is recovered

polyfractal opened the pull request:
If a bulk request hits a node that has not recovered the CS yet, index metadata will not be available. The actual bulk request will wait/retry until the CS is recovered (due to retry mechanism in `BulkOperation`), but the resolution of index templates has no retry mechanism.  This means `default-pipeline` processors will be omitted and this can result in incorrect documents being indexed.

This commit adds a blocking/retry mechanism so that index metadata resolution waits until the CS has been recovered.

## Questions/concerns
- I'm assuming we want to retry if the CS isn't ready... is that a correct assumption? 

- This follows a similar pattern to `BulkOperation`/`executeBulk()`, since it seems the TransportBulkAction is reused (at least in the integ tests) so we need to store state somewhere other than on the action itself.  

  It does makes `BulkOperation` somewhat redundant, since that also does cluster state checking.  But I was hesitant to change more than was necessary.  If desired I can try to meld the two together so there is less redundancy.

- Not familiar with this code, so any/all suggestions welcome :)

DaveCTurner commented on server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java:
```diff
@@ -147,149 +147,208 @@ public static IndexRequest getIndexWriteRequest(DocWriteRequest docWriteRequest)
+            if (startTime == -1) {
+               startTime = relativeTime();
+            }
+            ClusterBlockException blockException = clusterService.state().blocks().globalBlockedException(ClusterBlockLevel.METADATA_READ);
```
I think that `ClusterBlockLevel.WRITE` is more appropriate, since writing data is what we're trying to do.

DaveCTurner commented on server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java:
```diff
@@ -147,149 +147,208 @@ public static IndexRequest getIndexWriteRequest(DocWriteRequest docWriteRequest)
+            prepForBulk();
+        }
+
+        private void prepForBulk() {
```
I think we should avoid getting a fresh cluster state in this method and instead use the one we just verified to be free of any unwanted blocks, i.e. that was passed into `onNewClusterState`.

DaveCTurner commented on modules/ingest-common/src/test/java/org/elasticsearch/ingest/common/IngestRestartIT.java:
```diff
@@ -177,6 +186,98 @@ public void testPipelineWithScriptProcessorThatHasStoredScript() throws Exceptio
+
+        // so the cluster state can be recovered
+        internalCluster()
+            .startNode(Settings.builder().put(GatewayService.RECOVER_AFTER_NODES_SETTING.getKey(), "1"));
```
I think this can just be `.startNode()` - the other node is the master, so this node doesn't do any state recovery.

DaveCTurner commented on server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java:
```diff
@@ -147,149 +147,208 @@ public static IndexRequest getIndexWriteRequest(DocWriteRequest docWriteRequest)
+                    listener.onFailure(blockException);
+                    return;
+                }
+                recoveredObserver.waitForNextChange(new ClusterStateObserver.Listener() {
```
I think we could use `waitForNextChange(Listener listener, Predicate<ClusterState> statePredicate)` to filter out any intermediate states that still have a block - that way we don't need to check that again and can go straight into `prepForBulk`

DaveCTurner commented on server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java:
```diff
@@ -147,149 +147,208 @@ public static IndexRequest getIndexWriteRequest(DocWriteRequest docWriteRequest)
+                    @Override
+                    public void onTimeout(TimeValue timeout) {
+                        // Try one more time...
+                        run();
```
I think it's best just to fail here.

polyfractal commented:
Thanks @DaveCTurner!  Addressed review comments, the predicate definitely made it cleaner.  I _think_ everything is constructed correctly, but the global block methods/semantics confuse me a little so extra eyeballs there might be warranted :)

I also moved the start time into the ctor because it was simpler, and should probably be run right after instantiation anyway.

DaveCTurner commented on server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java:
```diff
@@ -147,149 +147,202 @@ public static IndexRequest getIndexWriteRequest(DocWriteRequest docWriteRequest)
+     * A runnable that will ensure the cluster state has been recovered enough to
+     * read index metadata and templates/pipelines.  Will retry up to the bulk's timeout
+     */
+    private final class BulkExecutor extends ActionRunnable<BulkResponse> {
```
Do we need this class now that we aren't doing any retries ourself? I get that it wraps up the three parameters and the start time but 4 parameters to another method isn't so bad.

DaveCTurner commented on server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java:
```diff
@@ -147,149 +147,202 @@ public static IndexRequest getIndexWriteRequest(DocWriteRequest docWriteRequest)
+
+        @Override
+        protected void doRun() {
+            ClusterState currentState = clusterService.state();
```
Technically I think we should get the state from `recoveredObserver.setAndGetObservedState()` to keep the state we observe in sync with the state that the `ClusterStateObserver` is observing. Alternatively, we could avoid constructing the `ClusterStateObserver` until we discover we need it, and give it the state we got from the cluster service - there's no need for it to be tracked in a field.

DaveCTurner commented on server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java:
```diff
@@ -147,149 +147,202 @@ public static IndexRequest getIndexWriteRequest(DocWriteRequest docWriteRequest)
+                    @Override
+                    public void onNewClusterState(ClusterState newState) {
+                        // predicate passed, begin preparing for the bulk
+                        prepForBulk(newState);
```
This is now sometimes on the cluster applier thread, which is probably not what we want. We should add this to the top of `prepForBulk`, and then make sure to call `prepForBulk` on the `WRITE` threadpool here.

```
assert ClusterApplierService.assertNotClusterStateUpdateThread("TransportBulkAction#prepForBulk");
```

polyfractal commented:
Urgh, this PR has languished.  I just don't have time to dig in and solve these issues I'm afraid, given my relative lack of experience in this part of the code.  When I started to fix the tests at the time, it became clear that the code needed a bit more adjusting than just fixing fragile mocks (e.g. to fix some of the mocks would have required pulling in half the world or even more excessive mocking).

I'm going to close this for now and open a ticket about the issue so it isn't lost.  I'll revisit this PR if I find some time but for now it's probably best to admit defeat :(
